sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala

```diff
@@ -123,7 +128,17 @@ object StreamingQueryListener {
       val id: UUID,
       val runId: UUID,
       val name: String,
-      val timestamp: String) extends Event
+      val timestamp: String) extends Event {
+
+    def json: String = compact(render(jsonValue))
+
+    private def jsonValue: JValue = {
```

Comment:
Discussion: Can we use Jackson API for the newly added code, for example:

```scala
    def json: String = jsonNode.toString

    private def jsonNode: JsonNode = {
      val node = JsonNodeFactory.instance.objectNode()
      node.put("id", id.toString)
      node.put("runId", runId.toString)
      node.put("name", name)
      node.put("timestamp", timestamp)
      node
    }
```

Reply:
Thanks for the reply! Searching for the usage of this API I actually found that there are already JSON serialization implemented for these events

https://github.com/apache/spark/blame/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryListenerSuite.scala#L267-L278 

I think I can actually remove these fields, wdyt? Also cc @HyukjinKwon @HeartSaVioR

Reply:
The test code you are referring to does not cover the case for version compatibility. JsonProtocol is actually purposed to handle backward-and-forward compatibility for events, but looks like we don't do it for SQL related events since JsonProtocol does not have a visibility for SQL module. Maybe we need to have another one for SQL when we figure out necessity of ensuring backward-and-forward compatibility.

That's the context. Regarding this change, even if we use Jackson here, underlying is still using json4s (see how StreamingQueryProgress is covered) hence it becomes a mix-up. I'd like to see the entire thing to be replaced with Jackson if we want to do that, for consistency. 

Could we allow json4s be used for this PR, and file a follow up JIRA ticket to replace entire part of streaming query listener events to use Jackson?

Reply:
I am fine to use json4s first in this one:)

Reply:
@HeartSaVioR Actually, a few days ago, I wrote a draft to use jackson api instead of json4s for `StreamingQueryProgress`： https://github.com/apache/spark/pull/41562 

However, maybe this is more suitable for change in Spark 4.0.

Reply:
I see. Thanks guys! Let's keep using json4s for this one and replace them all in a followup PR. @LuciferYang Just curious, I'm not very familiar with these two, is there a special benefit of switching to Jackson API?

Reply:
- https://github.com/apache/spark/pull/36885
- https://issues.apache.org/jira/browse/SPARK-39658

A old story, hope the above two links will be helpful to you :)

Reply:
I'm not too concerned about the performance for this (this is happening at most once per every microbatch execution), but I'd agree with the movement on removing json4s as a dependency, unless we decide to shade it.